File: WaterWaveDemo/WaterWaveDemo/SPWaterProgressIndicatorView.m
//
//  SPWaterProgressIndicatorView.m
//
//  Created by Antonio081014 on 9/4/15.
//  Copyright (c) 2015 antonio081014.com. All rights reserved.
//

#import "SPWaterProgressIndicatorView.h"

static const CGFloat kDefaultFrequency          = 1.5f;
static const CGFloat kDefaultAmplitude          = 1.0f;
static const CGFloat kDefaultIdleAmplitude      = 0.01f;
static const CGFloat kDefaultPhaseShift         = -0.15f;
static const CGFloat kDefaultDensity            = 5.0f;
static const NSUInteger kDefaultPercent         = 0;
static const NSUInteger kDefaultMaximumPercent  = 100;
static const CGFloat kDefaultPrimaryLineWidth   = 3.0f;
static const CGFloat kDefaultFontSize           = 20.f;
static const CGFloat kDefaultMinimumFontSize    = 12.f;

@interface SPWaterProgressIndicatorView ()
/*
 * Line width used for the proeminent wave
 * Default: 3.0f
 */
@property (nonatomic)  IBInspectable CGFloat primaryWaveLineWidth;

/*
 * The amplitude that is used when the incoming amplitude is near zero.
 * Setting a value greater 0 provides a more vivid visualization.
 * Default: 0.01
 */
@property (nonatomic) IBInspectable CGFloat idleAmplitude;

/*
 * The frequency of the sinus wave. The higher the value, the more sinus wave peaks you will have.
 * Default: 1.5
 */
@property (nonatomic) IBInspectable CGFloat frequency;

/*
 * The current amplitude
 */
@property (nonatomic) IBInspectable CGFloat amplitude;

/*
 * The lines are joined stepwise, the more dense you draw, the more CPU power is used.
 * Default: 5
 */
@property (nonatomic) IBInspectable CGFloat density;

/*
 * The phase shift that will be applied with each level setting
 * Change this to modify the animation speed or direction
 * Default: -0.15
 */
@property (nonatomic) IBInspectable CGFloat phaseShift;

@property (nonatomic) CGFloat phase;

@property (nonatomic) NSUInteger completionInPercent;

@property (nonatomic, strong) CADisplayLink *displaylink;

@property (nonatomic, strong) CATextLayer *textLayer;
@end

@implementation SPWaterProgressIndicatorView

- (instancetype)initWithFrame:(CGRect)frame
{
    if (self = [super initWithFrame:frame]) {
        [self setup];
    }
    
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder
{
    if (self = [super initWithCoder:aDecoder]) {
        [self setup];
    }
    
    return self;
}

- (void)setup
{
    // Default background color is transparent.
    self.backgroundColor = [UIColor clearColor];
    self.waveColor = [UIColor cyanColor];
    
    self.completionInPercent = kDefaultPercent;
    
    self.frequency = kDefaultFrequency;
    
    self.amplitude = kDefaultAmplitude;
    self.idleAmplitude = kDefaultIdleAmplitude;
    
    self.phaseShift = kDefaultPhaseShift;
    self.density = kDefaultDensity;
    
    self.primaryWaveLineWidth = kDefaultPrimaryLineWidth;
    self.fontSize = kDefaultFontSize;
    
    UIBezierPath *progressPath = [UIBezierPath bezierPathWithOvalInRect:self.bounds];
    CAShapeLayer *layer = [CAShapeLayer layer];
    [layer setPath:progressPath.CGPath];
    layer.lineWidth = MIN(CGRectGetHeight(self.bounds), CGRectGetWidth(self.bounds)) / 10.f;
    self.layer.mask = layer;
    
    [self.layer addSublayer:self.textLayer];
    
    [self startAnimation];
}

/**
 * Ensure the size is always square.
 */
- (void)setFrame:(CGRect)frame {
    if (frame.size.width < frame.size.height) {
        frame.size.height = frame.size.width;
    } else {
        frame.size.width = frame.size.height;
    }
    
    [super setFrame:frame];
}


- (CATextLayer *)textLayer
{
    if (!_textLayer) {
        _textLayer = [CATextLayer layer];
        _textLayer.bounds = CGRectMake(0, 0, CGRectGetWidth(self.bounds), MIN(CGRectGetHeight(self.bounds), 40.f));
        _textLayer.position = self.center;
        _textLayer.alignmentMode = kCAAlignmentCenter;
        _textLayer.string = [NSString stringWithFormat:@"%zd %%", self.completionInPercent];
        _textLayer.fontSize = self.fontSize;
        _textLayer.foregroundColor = [UIColor grayColor].CGColor;
    }
    return _textLayer;
}

- (void)setFontSize:(CGFloat)fontSize
{
    _fontSize = MAX(fontSize, kDefaultMinimumFontSize);
    
    [self setNeedsDisplay];
}

- (void)updateWithPercentCompletion:(NSUInteger)percent
{
    self.completionInPercent = MAX(0, MIN(kDefaultMaximumPercent, percent));
    
    self.textLayer.string = [NSString stringWithFormat:@"%zd %%", self.completionInPercent];
    
    [self setNeedsDisplay];
}

/**
 * With phase shifts, the animation will prevail.
 */
- (void)update
{
    self.phase += self.phaseShift;
    
    [self setNeedsDisplay];
}

- (void)startAnimation
{
    self.displaylink = [CADisplayLink displayLinkWithTarget:self selector:@selector(updateMeters)];
    [self.displaylink addToRunLoop:[NSRunLoop currentRunLoop] forMode:NSRunLoopCommonModes];
}

- (void)updateMeters
{
    [self update];
}

- (void)stopAnimation
{
    [self.displaylink removeFromRunLoop:[NSRunLoop currentRunLoop] forMode:NSRunLoopCommonModes];
    [self.displaylink invalidate];
    self.displaylink = nil;
}

//  Great Thanks to Stefan Ceriu.
//  Ref: https://github.com/stefanceriu/SCSiriWaveformView
- (void)drawRect:(CGRect)rect
{
    // Get current Context.
    CGContextRef context = UIGraphicsGetCurrentContext();
    // Clear everything in the bounds drawed in the last phase,
    // otherwise, drawing will overlap on each other.
    CGContextClearRect(context, self.bounds);
    
    [self.backgroundColor set];
    CGContextFillRect(context, rect);
    
    CGContextSetLineWidth(context, self.primaryWaveLineWidth);
    
    CGFloat halfHeight = CGRectGetHeight(self.bounds) / 2.0f;
    CGFloat width = CGRectGetWidth(self.bounds);
    CGFloat mid = width / 2.0f;
    
    const CGFloat maxAmplitude = fmax(halfHeight / 10 - 4.0f, 2.f * self.primaryWaveLineWidth); // 4 corresponds to twice the stroke width
    
    [[self.waveColor colorWithAlphaComponent:CGColorGetAlpha(self.waveColor.CGColor)] set];
    
    for (CGFloat x = 0; x<width + self.density; x += self.density) {
        // We use a parable to scale the sinus wave, that has its peak in the middle of the view.
        CGFloat scaling = -pow(1 / mid * (x - mid), 2) + 1;
        
        CGFloat y = scaling * maxAmplitude * self.amplitude * sinf(2 * M_PI *(x / width) * self.frequency + self.phase) + 1.0 * CGRectGetHeight(self.bounds) * (100 - self.completionInPercent) / 100.f;
        
        if (x == 0) {
            CGContextMoveToPoint(context, x, y);
        } else {
            CGContextAddLineToPoint(context, x, y);
        }
    }
    
    CGContextAddLineToPoint(context, width, CGRectGetHeight(self.bounds));
    CGContextAddLineToPoint(context, 0, CGRectGetHeight(self.bounds));
    CGContextClosePath(context);
    CGContextFillPath(context);
    CGContextStrokePath(context);
}

/**
 * Ensure nothing would happen when user touch out of the mask(circle)
 */
- (UIView *)hitTest:(CGPoint)point withEvent:(UIEvent *)event
{
    CAShapeLayer *layer = (CAShapeLayer *)self.layer.mask;
    if ([[UIBezierPath bezierPathWithCGPath:layer.path] containsPoint:point]){
        return self;
    }
    return nil;
}

@end

File: WaterWaveDemo_Swift/WaterWaveDemo_Swift/SPWaterProgressIndicatorView.swift
//
//  SPWaterProgressIndicatorView.swift
//
//  Created by Antonio081014 on 9/18/16.
//  Copyright © 2016 Antonio081014.com. All rights reserved.
//

import UIKit

class SPWaterProgressIndicatorView: UIView {
    
    var fontSize: CGFloat {
        didSet {
            self.setNeedsDisplay()
        }
    }
    var waveColor: UIColor {
        didSet {
            self.setNeedsDisplay()
        }
    }
    
    var completionInPercent: Int {
        didSet {
            if completionInPercent > 100 {
                completionInPercent = 100
            } else if completionInPercent < 0 {
                completionInPercent = 0
            }
            self.textLayer.string = String(format: "%zd %%", self.completionInPercent)
            self.setNeedsDisplay()
        }
    }
    
    
    func startAnimation() {
        self.displaylink = CADisplayLink(target: self, selector: #selector(SPWaterProgressIndicatorView.updateMeters))
        self.displaylink?.add(to: RunLoop.main, forMode: RunLoop.Mode.common)
    }
    
    func stopAnimation() {
        self.displaylink?.remove(from: RunLoop.current, forMode: RunLoop.Mode.common)
        self.displaylink?.invalidate()
        self.displaylink = nil
    }
    
    convenience init() {
        self.init(frame: .zero)
    }
    
    /// Designate Initializer
    override init(frame: CGRect) {
        self.waveColor = UIColor.cyan
        
        self.completionInPercent = kDefaultPercent
        
        self.frequency = kDefaultFrequency
        
        self.amplitude = kDefaultAmplitude
        self.idleAmplitude = kDefaultIdleAmplitude
        
        self.phaseShift = kDefaultPhaseShift
        self.density = kDefaultDensity
        
        self.primaryWaveLineWidth = kDefaultPrimaryLineWidth
        self.fontSize = kDefaultFontSize
        self.phase = 0
        self.displaylink = CADisplayLink.init()
        
        super.init(frame: frame)
        
        let progressPath = UIBezierPath(ovalIn: self.bounds)
        let layer = CAShapeLayer.init()
        layer.path = progressPath.cgPath
        layer.lineWidth = min(self.bounds.height, self.bounds.width) / 10.0
        self.layer.mask = layer
        
        self.layer.addSublayer(self.textLayer)
        
        self.backgroundColor = UIColor.clear
        
        self.startAnimation()
    }
    
    required convenience init?(coder aDecoder: NSCoder) {
        self.init(frame: .zero)
        fatalError("init(coder:) has not been implemented")
    }
    
    /**
     * Ensure the size is always square.
     */
    override var frame: CGRect {
        didSet {
            if frame.width != frame.height {
                let side = min(frame.width, frame.height)
                frame = CGRect.init(x: frame.origin.x, y: frame.origin.y, width: side, height: side)
            }
        }
    }
    
    let kDefaultFrequency: CGFloat = 1.5
    let kDefaultAmplitude: CGFloat = 1.0
    let kDefaultIdleAmplitude: CGFloat = 0.01
    let kDefaultPhaseShift: CGFloat = -0.15
    let kDefaultDensity: CGFloat = 5.0
    let kDefaultPercent: Int = 0
    let kDefaultMaximumPercent: Int = 100
    let kDefaultPrimaryLineWidth: CGFloat = 3.0
    let kDefaultFontSize: CGFloat = 20.0
    let kDefaultMinimumFontSize: CGFloat = 12.0
    
    /**
     * Line width used for the proeminent wave
     *
     * Default: 3.0f
     */
    fileprivate var primaryWaveLineWidth: CGFloat
    
    /**
     * The amplitude that is used when the incoming amplitude is near zero.
     * Setting a value greater 0 provides a more vivid visualization.
     *
     * Default: 0.01
     */
    fileprivate var idleAmplitude: CGFloat
    
    /**
     * The frequency of the sinus wave. The higher the value, the more sinus wave peaks you will have.
     *
     * Default: 1.5
     */
    fileprivate var frequency: CGFloat
    
    /**
     * The current amplitude
     */
    fileprivate var amplitude: CGFloat
    
    /**
     * The lines are joined stepwise, the more dense you draw, the more CPU power is used.
     *
     * Default: 5
     */
    fileprivate var density: CGFloat
    
    /**
     * The phase shift that will be applied with each level setting
     * Change this to modify the animation speed or direction
     *
     * Default: -0.15
     */
    fileprivate var phaseShift: CGFloat
    fileprivate var phase: CGFloat
    
    fileprivate var displaylink: CADisplayLink?
    fileprivate lazy var textLayer: CATextLayer = {
        let textlayer = CATextLayer.init()
        textlayer.bounds = CGRect(x: 0, y: 0, width: self.bounds.width, height: min(self.bounds.height, 40.0))
        textlayer.position = self.center
        textlayer.alignmentMode = CATextLayerAlignmentMode.center
        textlayer.string = String.init(format: "%zd %%", self.completionInPercent)
        textlayer.fontSize = CGFloat(self.fontSize)
        textlayer.foregroundColor = UIColor.gray.cgColor
        return textlayer
    }()
    
    @objc func updateMeters() {
        self.update()
    }
    
    /**
     * With phase shifts, the animation will prevail.
     */
    fileprivate func update() {
        self.phase += self.phaseShift
        self.setNeedsDisplay()
    }
    
    override func draw(_ rect: CGRect) {
        // Get current Context.
        let context = UIGraphicsGetCurrentContext()
        
        // Clear everything in the bounds drawed in the last phase,
        // otherwise, drawing will overlap on each other.
        context?.clear(self.bounds)
        
        self.backgroundColor?.set()
        
        context?.fill(rect)
        context?.setLineWidth(CGFloat(self.primaryWaveLineWidth))
        
        let halfHeight = self.bounds.height / 2.0
        let width: CGFloat = self.bounds.width
        let mid = width / 2.0
        
        let maxAmplitude = max(halfHeight / 10 - 4.0, CGFloat(2.0 * self.primaryWaveLineWidth)) // 4 corresponds to twice the stroke width
        
        self.waveColor.withAlphaComponent(self.waveColor.cgColor.alpha).set()
        
        var x = CGFloat(0)
        while  x < (width + self.density) {
            // We use a parable to scale the sinus wave, that has its peak in the middle of the view.
            let scaling = -pow(1 / mid * (x - mid), 2) + 1
            
            let y = scaling * maxAmplitude * self.amplitude * sin(CGFloat(2.0 * CGFloat.pi) * (x / width) * self.frequency + self.phase) + self.bounds.height * CGFloat(100 - self.completionInPercent) / CGFloat(100)
            
            if (x == 0) {
                context?.move(to: CGPoint.init(x: x, y: y))
            } else {
                context?.addLine(to: CGPoint.init(x: x, y: y))
            }
            
            x += self.density
        }
        context?.addLine(to: CGPoint.init(x: width, y: self.bounds.height))
        context?.addLine(to: CGPoint.init(x: CGFloat(0), y: self.bounds.height))
        context?.closePath()
        context?.fillPath()
        context?.strokePath()
    }
    
    
    /**
     * Ensure nothing would happen when user touch out of the mask(circle)
     */
    override func hitTest(_ point: CGPoint, with event: UIEvent?) -> UIView? {
        if let layer = self.layer.mask as? CAShapeLayer {
            if UIBezierPath.init(cgPath: layer.path!).contains(point) {
                return super.hitTest(point, with: event)
            }
            return nil
        }
        return nil
    }
}
